luthier/src/main/java/com/yahoo/luthier/webservice/data/config/metric/MetricMakerDictionary.java

```diff
@@ -0,0 +1,213 @@
+                }
+
+                assert selectedConstructor != null;
+                Class<?>[] paramsTypes = selectedConstructor.getParameterTypes();
+                String[] paramsNames = discoverer.getParameterNames(selectedConstructor);
+
+                Object[] args = IntStream.range(0, paramsTypes.length)
+                        .mapToObj(i -> parseParams(paramsTypes[i].getSimpleName(), paramsNames[i], maker))
+                        .toArray();
+
+                add(maker.getName(), (MetricMaker) selectedConstructor.newInstance(args));
+
+            } catch (ClassNotFoundException | IllegalAccessException | InstantiationException e) {
+                LOG.error("The constructor of maker's class is inaccessible", e);
+            } catch (InvocationTargetException e) {
+                LOG.error(e.getCause().getMessage(), e);
+            }
+        });
+    }
+
+    /**
+     * Find a metric maker given a metric maker Name.
+     *
+     * @param metricMakerName Name to search
+     * @return the first metric maker found (if exists)
+     */
+    public MetricMaker findByName(String metricMakerName) {
+        return nameToMetricMaker.get(metricMakerName.toLowerCase(Locale.ENGLISH));
+    }
+
+    /**
+     * Get all metric makers available in metric maker dictionary.
+     *
+     * @return a set of metric makers
+     */
+    public Set<MetricMaker> findAll() {
+        return Collections.unmodifiableSet(new HashSet<>(nameToMetricMaker.values()));
+    }
+
+    /**
+     * Adds the specified element to the dictionary if it is not already present.
+     *
+     * @param name        key to add to dictionary
+     * @param metricMaker value to add to dictionary
+     * @return <tt>true</tt> if the dictionary did not already contain the specified metric maker
+     */
+    public boolean add(String name, MetricMaker metricMaker) {
+        if (nameToMetricMaker.containsKey(name.toLowerCase(Locale.ENGLISH))) {
+            return false;
+        }
+        MetricMaker metricMakers = nameToMetricMaker.put(name.toLowerCase(Locale.ENGLISH), metricMaker);
+        if (metricMakers != null) {
+            // should never happen unless multiple loaders are running in race-condition
+            ConcurrentModificationException e = new ConcurrentModificationException();
+            LOG.error("Multiple loaders updating MetricMakerDictionary", e);
+            throw e;
+        }
+        return true;
+    }
+
+    /**
+     * Adds all of the metric makers in the specified collection to the dictionary.
+     *
+     * @param metricMakers collection of metric makers to add
+     * @return <tt>true</tt> if the dictionary changed as a result of the call
+     */
+    public boolean addAll(Collection<MetricMaker> metricMakers) {
+        boolean flag = false;
+        for (MetricMaker metricMaker : metricMakers) {
+            flag = add(metricMaker.toString(), metricMaker) || flag;
+        }
+        return flag;
+    }
```

Comment:
you might want to short circuit here and throw an error. 

@michael-mclawhorn what do you think?